extensions-contrib/kafka-emitter/src/main/java/org/apache/druid/emitter/kafka/KafkaEmitter.java

```diff
@@ -197,10 +236,43 @@ public void emit(final Event event)
     }
   }
 
+  private void emitSegmentMetadataEvent(SegmentMetadataEvent event)
+  {
+    Map<String, Object> eventMap = event.toMap();
+    long createdTimeMs = ((DateTime) eventMap.get(SegmentMetadataEvent.CREATED_TIME)).getMillis();
+    Timestamp createdTimeTs = Timestamp.newBuilder().setSeconds(createdTimeMs / 1000)
+        .setNanos((int) ((createdTimeMs % 1000) * 1000000)).build();
+    long startTimeMs = ((DateTime) eventMap.get(SegmentMetadataEvent.START_TIME)).getMillis();
+    Timestamp startTimeTs = Timestamp.newBuilder().setSeconds(startTimeMs / 1000)
+        .setNanos((int) ((startTimeMs % 1000) * 1000000)).build();
+    long endTimeMs = ((DateTime) eventMap.get(SegmentMetadataEvent.END_TIME)).getMillis();
+    Timestamp endTimeTs = Timestamp.newBuilder().setSeconds(endTimeMs / 1000)
+        .setNanos((int) ((endTimeMs % 1000) * 1000000)).build();
+
+    DruidSegmentEvent.Builder druidSegmentEventBuilder = DruidSegmentEvent.newBuilder()
+        .setDatasource(eventMap.get(SegmentMetadataEvent.DATASOURCE).toString())
+        .setCreatedTime(createdTimeTs)
+        .setStartTime(startTimeTs)
+        .setEndTime(endTimeTs)
+        .setVersion(eventMap.get(SegmentMetadataEvent.VERSION).toString())
+        .setIsCompacted((boolean) (eventMap.get(SegmentMetadataEvent.IS_COMPACTED)));
+    if (config.getClusterName() != null) {
+      druidSegmentEventBuilder.setClusterName(config.getClusterName());
+    }
+    DruidSegmentEvent druidSegmentEvent = druidSegmentEventBuilder.build();
+    try {
+      metadataProducer.send(new ProducerRecord<>(config.getMetadataTopic(), druidSegmentEvent), setProducerCallback(segmentMetadataLost));
```

Comment:
If we look at emitter interface it states the following

/**
   * Emit an event. This method must not throw exceptions or block.
   *
   * If an implementation receives too many events and internal queues fill up, it should drop events rather than
   * blocking or consuming excessive memory.
   *
   * If an implementation receives input it considers to be invalid, or has an internal problem, it should deal with
   * that by logging a warning rather than throwing an exception. Implementations that log warnings should consider
   * throttling warnings to avoid excessive logs, since a busy Druid cluster can emit a high volume of metric events.
   */

I also see some article that producer send api can block in few scenarios

https://stackoverflow.com/questions/47357590/kafka-producer-send-blocks-indefinitely-when-kafka-servers-are-down#:~:text=Kafka%20producer%20send%20blocks%20indefinitely%20when%20kafka%20servers%20are%20down,-Ask%20Question

I think we should introduce MemoryBoundLinkedBlockingQueue even for this event type and handle exceptions and retries above the send api call. This way we make sure that our druid performance will never be affected.

Reply:
you are correct that it's possible for a producer to block, but I think we can address this separately, since this is not a new issue with KafkaEmitter, and it probably warrants its own PR and discussion.

Reply:
will come back to this after finalizing the approach [here](https://github.com/confluentinc/druid/pull/106/files#r971431312)

Reply:
>since this is not a new issue with KafkaEmitter

What do you mean by it's not a new issue?  Currently, we use `scheduler.schedule` to schedule emitting of other events from Kafka topic. Does the scheduler run them in a sperate thread? If yes, then they shouldn't block the main thread execution when Kafka is unreachable, etc right?

Reply:
True I forgot there was a separate queue. We could leverage the same queue the emitter already has, and maybe increase the frequency of collecting these events. Eventually we should move this emitter to send data continuously from the queue rather than on a schedule so we don't introduce artificial delay unless there is actually a problem delivering events.

Reply:
The emitter actually emits these events continuously. There's only a one-time delay of 10s at the start.

Reply:
@harinirajendran good point about sending events continuously. I think using the same mechanisms makes complete sense in that case, no need to create a different path to send events.